test.rb

```diff
@@ -0,0 +1,176 @@
+      # forks and inherits stdout, as happens in `foo &`.
+      until results = Process.waitpid2(child_pid, Process::WNOHANG)
+
+        # If not done, wait for data to read with a timeout
+        # This timeout is selected to keep activity low while waiting on
+        # a long process, while not waiting too long for the pathological
+        # case where stdout is never closed.
+        ready = IO.select([reader], [], [], 0.1)
+        begin
+          output << reader.read_nonblock(4096) if ready
+        rescue Errno::EAGAIN
+        rescue EOFError
+        end
+      end
+
+      # Read any remaining data. Allow for but don't expect EOF.
+      begin
+        loop do
+          output << reader.read_nonblock(4096)
+        end
+      rescue Errno::EAGAIN
+      rescue EOFError
+      end
+
+      # Force to external encoding to preserve prior behavior when reading a file.
+      # Wait until after reading all data so we don't encounter corruption when
+      # reading part of a multi-byte unicode character if default_external is UTF-8.
+      output.force_encoding(Encoding.default_external)
+      exit_status = results.last.exitstatus
+      child_pid = nil
+    rescue Timeout::Error => e
+      # NOTE: For Ruby 2.1+, an explicit Timeout::Error class has to be
+      # passed to Timeout.timeout in order for there to be something for
+      # this block to rescue.
+      unless child_pid.nil?
+        Process.kill(:TERM, child_pid)
+        # Spawn a thread to reap the process if it dies.
+        Thread.new { Process.waitpid(child_pid) }
+      end
+
+      raise e
+    end
+  ensure
+    # Make sure all handles are closed in case an exception was thrown attempting to execute.
+    [stdin, stdout, stderr].each do |io|
+      io.close
+    rescue StandardError
+      nil
+    end
+  end
+
+  output
+end
+
+
+## Load bolt and get an application instance we can use to run plans in a fork
+project = Bolt::Project.find_boltdir(Dir.pwd)
+config = Bolt::Config.from_project(project, {})
+analytics = Bolt::Analytics::NoopClient.new
+executor = Bolt::Executor.new(
+  config.concurrency,
+  analytics,
+  false,
+  config.modified_concurrency,
+  config.future
+)
+pal = Bolt::PAL.new(
+  Bolt::Config::Modulepath.new(config.modulepath),
+  config.hiera_config,
+  config.project.resource_types,
+  config.compile_concurrency,
+  config.trusted_external,
+  config.apply_settings,
+  config.project
+)
+plugins = Bolt::Plugin.new(config, pal, analytics)
+inventory = Bolt::Inventory.from_config(config, plugins)
+application = Bolt::Application.new(
+  analytics: analytics,
+  config: config,
+  executor: executor,
+  inventory: inventory,
+  pal: pal,
+  plugins: plugins
+)
+
+
+
+
+# TODO: experiment with runnign these concurrently (not currently working serially)
+# results = 10.times.map { execute_plan(application, 'bolt_test') }
+# puts results
+
+
+
+report = MemoryProfiler.report do
+  threads = []
+  10.times {
```

Comment:
TODO: 
We can use the pattern here for creating an `Application` object to load a project and then spin off plans in forks from there. This was the part that required the most "bolt specific" knowledge. 

Instead of the pattern here of threading then forking (and waiting for a result) we will instead:

Fork N plans at a time and wait. Rob and Austin will take this on.